### DISCUSSION TITLE
What is the role of Binary?

### DISCUSSION THREAD

**Discussion by Xazax-hun** [GabrielDosReis/ipr#261] [Q&A] [answered]
After https://github.com/GabrielDosReis/ipr/commit/802c786418ab9104d9e348756cc36487f7ad059f, `Handler` is no longer Binary, but its interface still has only two getters. We had some code that relied on some nodes being `Binary` so we can reduce the amount of code we need to write by utilizing the more regular structure of IPR. Does this mean we should no longer rely on nodes being `Binary` or `Unary` but call the specific getters directly? (Instead of using first/second in generic code.)

**Comment by GabrielDosReis** [comment, marked as answer]
The structural classes `Unary`, `Binary`, and `Ternary` are devices to capture structural commonalities of various IPR nodes in order to reduce implementation complexities (see how their are used `<ipr/impl>`).  Their accessor functions are non-descriptive names such as `first()`, `second()`, and `third()` and -- in general -- not reflective of a deeper meaning. In particular, any permutation of such accessors is a structural isomorphism.  Consequently, they are not interface functions to rely on directly.  Of course, there are exceptions, e.g. arithmetic expressions, where a `first() or `second()` operand have more interface meaning.

If there is a more actionable issue to address in form of code, please feel free to create an issue for that.

**Comment by Xazax-hun** [comment]
>  Consequently, they are not interface functions to rely on directly.

Why? There are many scenarios where we do not care about the order of the constituents of a node. E.g., when we pretty print a node. In those scenarios, we are not interested in the deeper meaning of the children, we only want to make sure we visit everything we are interested in. No longer capturing these commonalities for some nodes can increase the complexity of some user code as we will no longer be able to use our generic Binary printer for `Handler`, we will need to introduce our own, which will ultimately end up doing the exact same thing but using methods with different names. This just forces us to write more boilerplate code without adding any substance. 

Moreover, IPR gives little to no guidance on how it should be used/created/consumed. These methods should not have been relied on were not clearly communicated anywhere.

I am ok with IPR not considering these part of the interface, but in that case I believe IPR should provide the facilities to traverse the graph.

**Reply by GabrielDosReis** [reply]
> Why?

I just explained the reasoning, hence the word _Consequently_.  

> here are many scenarios where we do not care about the order of the constituents of a node. E.g., when we pretty print a node.

Actually, that example contradicts the assertion it is intended to illustrate: the pretty printing would want to avoid confusing the left operand of a `<<` with its right operand, for example.  The fact that you can  find _some_ context where the ordering does not matter does not necessarily imply all _for all_ use contexts, the ordering does not matter.

> Moreover, IPR gives little to no guidance on how it should be used/created/consumed. 

That is a fair observation, and the original question helps cover them.

**Reply by Xazax-hun** [reply]
> I just explained the reasoning, hence the word Consequently.

While admittedly I have a long way to go to improve my English knowledge, my clarifying question did not refer to your choice of words. I specifically was surprised about your reasoning as such statements seem to imply that implementation details are leaked into the interface nodes. With IPR's strong separation of interfaces and implementations and giving the option to roll your own implementation on top of the interfaces I was surprised to learn that certain implementation facilities manifest at the interface level, being prescriptive about how an implementation should look like. 

> Actually, that example contradicts the assertion it is intended to illustrate: the pretty printing would want to avoid confusing the left operand of a << with its right operand, for example. 

I don't see the contradiction. Pretty printing will always start with printing `first`, so there is no confusion in the output. In our Clang->IPR conversion tool, we created convenience facilities to traverse IPR, visiting every node once, and these facilities worked great for many purposes including pretty printing the representation or writing verifiers that check validates certain invariants that are not captured by the type system. `Binary` was useful to reduce the boilerplate code in those facilities.

> That is a fair observation, and the original question helps cover them.

Do we expect new users to read all conversations and issues? Or do you plan to keep answering the same questions all over again? Until it is actually documented in an accessible place, I do not consider the questions actually covered.

**Reply by GabrielDosReis** [reply]
> Do we expect new users to read all conversations and issues? Or do you plan to keep answering the same questions all over again? Until it is actually documented in an accessible place, I do not consider the questions actually covered.

One of the benefits of having dedicated **discussion (Q&A)** category of issues is precisely to provide a place for people to look up topics that may have been discussed in the past, so I don't have (or nobody else has) to answer the same question all over again.   Please make good use of it, as answers provided here live on for ever.  It is also a great way to have a forum to refine answers as we learn more or our collective thinking evolves -- as opposed to "closed" issues.  Treat this space as part of the documentation.  In addition to that, anyone should feel free to propose other form of documentation as PRs for consideration.  Existence of documentation has never prevented anyone from asking an already answered question (or a question already answered in a documentation).

**Reply by Xazax-hun** [reply]
> Treat this space as part of the documentation.

Sounds fair, but in this case it would be nice to call this out in the readme or some similar places as this is somewhat unique among the open source projects I worked with and I might not be the only one who would not start searching in the discussions to find this sort of information.

**Reply by GabrielDosReis** [reply]
Thanks for sharing your thoughts on how I could structure this repo.  I don't plan on calling out every tab in the README; that pops up in the GitHub interface as the larger community evolves  and develops new useful ways of building software.   Have a look at #187.

**Reply by Xazax-hun** [reply]
> I don't plan on calling out every tab in the README;

I was not referring to calling out the existence of the tab but calling out the nature of information that one is expected to find there. E.g., when a user is not supposed to use rely on a publicly available interface method, I would expect to see warnings in the source code rather than on some kind of discussion forums. But your repo, your way.

**Reply by GabrielDosReis** [reply]
Gabor -

You said:
> but in this case it would be nice to call this out in the readme or some similar places as this is somewhat unique among the open source projects I worked with and I might not be the only one who would not start searching in the discussions to find this sort of information.

If you have constructive comments to make to help improve the library, please feel free to share that with all of us.  Otherwise, I would ask you to exercise  more thoughts about whether you're moving things forward, or if you derailing forward progress.  Example of how to help move things forward was shown by @d-winsor  through his thoughtful summary and injection of background.

If further comments don't help advance understanding of the real issue, I will be left with no other choice but to lock this particular conversation (which will be real bummer).

**Reply by Xazax-hun** [reply]
> If you have constructive comments to make to help improve the library,

I thought the suggestion of documenting the fact whether and in what circumstances should a public interface function be relied on belongs to the source code rather than a forum was constructive. You disagreed. Ok. That was the end of the discussion on my end.

>  Example of how to help move things forward was shown by @d-winsor through his thoughtful summary and injection of background.

You specifically asked me in the past to state the problems and not to start with the solutions, yet later you praise someone who was summarizing possible solutions. I merely followed the guidelines you gave me in the past. 

But don't worry. I realized that we cannot get to the same page no matter what approach I am using. Anything I said must be wrong, not worthy of consideration. So I no longer plan to engage in any discussions regarding IPR.
